### DISCUSSION TITLE
redbpf / cargo bpf?

### DISCUSSION THREAD

**Discussion by djahandarie** [ipetkov/crane#36] [Q&A] [answered]
Absolutely love the project, this is one of the most ergonomic Rust nix flakes integrations I've seen so far.

I have a project where I'd like to use https://github.com/foniod/redbpf, which provides its own setup using a Cargo subcommand it provides called `cargo bpf`.

What would be the 'right' way to use this with crane?

**Comment by ipetkov** [comment, marked as answer]
Hi @djahandarie, glad you're enjoying Crane!

I didn't see `cargo-bpf` present in nixpkgs, so the first step would be to package it such that it is available to your project.

Once that is done, [cargoBuild](https://github.com/ipetkov/crane/blob/39305b3828b976cac90bcb2cc5bdf5cfbcafce33/docs/API.md#libcargobuild) is probably the wrapper you could use if you want to model that specific invocation as a derivation. It will handle vendoring the workspace dependencies, reuse workspace artifacts, etc., and then run whatever cargo command you'd like. Here's a short example:

```nix
let
  src = ./src;
  # Assuming you want to prebuild the workspace with regular
  # `cargo build`, or set it to null if you want to skip this step
  cargoArtifacts = crane.buildDepsOnly {
    inherit src;
  };
in
crane.cargoBuild {
  inherit src cargoArtifacts;

  cargoBuildCommand = "cargo bpf whatever args needed";
  nativeBuildInputs = [
    cargo-bpf
  ];
}
```

Does this help?
